src/Data/Voting.hs

```diff
@@ -0,0 +1,518 @@
+
+emptyDelegationState = DelegationState emptyDelegations emptyVotings emptyTopicTree
+
+setDelegationPure :: Voter -> Topic -> Voter -> Delegations -> Delegations
+setDelegationPure from topic to (Delegations dm@(DelegationMap dmap) (CoDelegationMap coDmap))
+    = Delegations dmap' coDmap'
+  where
+    mOldTo = lookupDoubleMap from topic dmap
+
+    dmap'   = DelegationMap (insertDoubleMap from topic to dmap)
+    coDmap' = CoDelegationMap coDmap1
+    coDmap0 = maybe coDmap (\to' -> coDmap & at to' . _Just . at topic . _Just . at from .~ Nothing
+                                           & at to' . _Just . at topic %~ deleteEmpty
+                                           & at to' %~ deleteEmpty) mOldTo
+    coDmap1 = insertDoubleMap to topic (Set.insert from $ fromMaybe Set.empty (lookupDoubleMap to topic coDmap0)) coDmap0
+
+    deleteEmpty (Just s) | null s = Nothing
+    deleteEmpty x                 = x
+
+
+-- | Non empty list of topics, from leaf to root.
+topicHiearchyPure :: Topic -> TopicTree -> [Topic]
+topicHiearchyPure t (TopicTree tmap) = fix path t
+  where
+    path rec t = maybe [t] ((t:) . rec) (Map.lookup t tmap)
+
+canVotePure :: Voter -> Idea -> DelegationMap -> TopicTree -> Bool
+canVotePure v i dmap ttree =
+    all isNothing $ (\t -> lookupDMap v t dmap) <$> topicHiearchyPure (TopicIdea i) ttree
+
+getSupportersPure :: Voter -> Topic -> DelegationMap -> CoDelegationMap -> TopicTree -> [Voter]
+getSupportersPure v t dmap (CoDelegationMap codmap) ttree = fix voters v
+  where
+    -- idea -> class -> school
+    topicPath  = topicHiearchyPure t ttree
+
+    -- If the first voter in the candidate list is the `v`, `v` is
+    -- responsible for the voting for the user u
+    supporter u =
+        let cm = candidates u dmap
+        in case catMaybes $ map (\t -> Map.lookup t cm) topicPath of
+            []     -> False
+            v' : _ -> v' == v
+
+    voters rec v =
+        v : maybe [] (mconcat . map rec)
+                     (filter supporter . Set.toList <$> lookupDoubleMap v t codmap)
+
+setVoteForPure :: Voter -> Idea -> Vote -> Voter -> Votings -> Votings
+setVoteForPure voter idea vote delegator (Votings vmap) =
+    Votings $ Map.insert (delegator, idea) (voter, vote) vmap
+
+setTopicDepPure :: Topic -> Topic -> TopicTree -> Either DelegationError TopicTree
+setTopicDepPure f t tt | elem f (topicHiearchyPure t tt) = Left DelegationCircularTopicDependency
+setTopicDepPure f t tm@(TopicTree tmap) =
+    Right $ case Map.lookup t tmap of
+        Just f' | f == f' -> tm -- Avoid circular deps should throw an error
+        _                 -> TopicTree (Map.insert f t tmap)
+
+getVotePure :: Voter -> Idea -> Votings -> Maybe (Voter, Vote)
+getVotePure v i (Votings vmap) = Map.lookup (v,i) vmap
+
+-- * monadic api
+
+class Monad m => DelegationM m where
+    setDelegation :: Voter -> Topic -> Voter -> m ()
+    canVote       :: Voter -> Idea  -> m Bool
+    getSupporters :: Voter -> Topic -> m [Voter]
```

Comment:
i guess we need to look at the way this called.  my expectation is that andor's version is more convenient, but i don't trust my expectation that much here.